src/realm/sync/apply_to_state_command.cpp

```diff
@@ -0,0 +1,407 @@
+                                        compressed_body_size, sv.size());
+        }
+
+        ret.uncompressed_body_buffer.set_size(uncompressed_body_size);
+        auto ec = realm::_impl::compression::decompress(sv.data(), compressed_body_size,
+                                                        ret.uncompressed_body_buffer.data(), uncompressed_body_size);
+        if (ec) {
+            throw MessageParseException("error decompressing message body: %1", ec.message());
+        }
+
+        ret.remaining = sv.substr(compressed_body_size);
+        ret.body_view = StringView{ret.uncompressed_body_buffer.data(), uncompressed_body_size};
+    }
+    else {
+        if (sv.size() < uncompressed_body_size) {
+            throw MessageParseException("message body is bigger (%1) than available bytes (%2)",
+                                        uncompressed_body_size, sv.size());
+        }
+        ret.body_view = sv.substr(0, uncompressed_body_size);
+        ret.remaining = sv.substr(uncompressed_body_size);
+    }
+
+    return ret;
+}
+
+ParseResult<Message> parse_message(StringView sv, Logger* logger)
+{
+    StringView message_type;
+    sv = parse_header_element(sv, ' ', message_type);
+
+    if (message_type == "download") {
+        return DownloadMessage::parse(sv, logger);
+    }
+    else if (message_type == "upload") {
+        return UploadMessage::parse(sv, logger);
+    }
+    else if (message_type == "ident") {
+        return ServerIdentMessage::parse(sv);
+    }
+    throw MessageParseException("could not find valid message in input");
+}
+
+ParseResult<ServerIdentMessage> ServerIdentMessage::parse(StringView sv)
+{
+    ServerIdentMessage ret;
+    sv = parse_header_line(sv, '\n', ret.session_ident, ret.file_ident.ident, ret.file_ident.salt);
+
+    return std::make_pair(std::move(ret), sv);
+}
+
+ParseResult<DownloadMessage> DownloadMessage::parse(StringView sv, Logger* logger)
+{
+    DownloadMessage ret;
+    int is_body_compressed;
+    std::size_t uncompressed_body_size, compressed_body_size;
+
+    sv = parse_header_line(sv, '\n', ret.session_ident, ret.progress.download.server_version,
+                           ret.progress.download.last_integrated_client_version, ret.latest_server_version.version,
+                           ret.latest_server_version.salt, ret.progress.upload.client_version,
+                           ret.progress.upload.last_integrated_server_version, ret.downloadable_bytes,
+                           is_body_compressed, uncompressed_body_size, compressed_body_size);
+
+    auto message_body = MessageBody::parse(sv, compressed_body_size, uncompressed_body_size, is_body_compressed);
+    ret.uncompressed_body_buffer = std::move(message_body.uncompressed_body_buffer);
+    sv = message_body.remaining;
+    auto body_view = message_body.body_view;
+
+    logger->trace("decoding download message. "
```

Comment:
If we are using the logger unconditionally, might as well take it as a reference parameter instead of a pointer.

Reply:
Done. I do dislike mutable references as function arguments though.